staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,802 @@
+    pub unstaked_available_epoch_height: EpochHeight,
+}
+
+impl Default for Account {
+    fn default() -> Self {
+        Self {
+            unstaked: 0,
+            stake_shares: 0,
+            unstaked_available_epoch_height: 0,
+        }
+    }
+}
+
+/// The number of epochs required for the locked balance to become unlocked.
+const NUM_EPOCHS_TO_UNLOCK: EpochHeight = 3;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StakingContract {
+    /// The account ID of the owner who's running the staking validator node.
+    /// NOTE: This is different from the current account ID which is used as a validator account.
+    /// The owner of the staking pool can change staking public key and adjust reward fees.
+    pub owner_id: AccountId,
+    /// The public key which is used for staking action. It's the public key of the validator node
+    /// that validates on behalf of the pool.
+    pub stake_public_key: PublicKey,
+    /// The last epoch height when `ping` was called.
+    pub last_epoch_height: EpochHeight,
+    /// The last known amount of locked balance of this account.
+    pub last_locked_account_balance: Balance,
+    /// The last known amount of unlocked balance of this account.
+    pub last_account_balance: Balance,
+    /// The total amount of shares. It should be equal to the total amount of shares across all
+    /// accounts.
+    pub total_stake_shares: Balance,
+    /// The total staked balance.
+    pub total_staked_balance: Balance,
+    /// The fraction of the reward that goes to the owner of the staking pool for running the
+    /// validator node.
+    pub reward_fee_fraction: RewardFeeFraction,
+    /// Persistent map from an account ID hash to the corresponding account.
+    pub accounts: Map<AccountHash, Account>,
+}
+
+/// Returns sha256 hash of the given account ID.
+fn hash_account_id(account_id: &AccountId) -> Vec<u8> {
+    env::sha256(account_id.as_bytes())
+}
+
+impl Default for StakingContract {
+    fn default() -> Self {
+        env::panic(b"Staking contract should be initialized before usage")
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone)]
+pub struct RewardFeeFraction {
+    pub numerator: u32,
+    pub denominator: u32,
+}
+
+impl RewardFeeFraction {
+    pub fn assert_valid(&self) {
+        assert_ne!(self.denominator, 0, "Denominator must be a positive number");
+        assert!(
+            self.numerator <= self.denominator,
+            "The reward fee must be less or equal to 1"
+        );
+    }
+
+    pub fn multiply(&self, value: Balance) -> Balance {
+        (U256::from(self.numerator) * U256::from(value) / U256::from(self.denominator)).as_u128()
+    }
+}
+
+/// Interface for a voting contract.
+#[ext_contract(ext_voting)]
+pub trait VoteContract {
+    /// Votes on the given proposal_id.
+    fn vote(&mut self, proposal_id: ProposalId);
```

Comment:
there should also be how much stake you vote

Reply:
Will add this